packages/transport-tcp/test/connection-limits.spec.ts

```diff
@@ -0,0 +1,217 @@
+
+    await assertConnectedSocket(1)
+    await assertConnectedSocket(2)
+    await assertConnectedSocket(3)
+    await assertServerConnections(listener, 3)
+
+    // Limit reached, server should be closed here
+    await assertRefusedSocket(4)
+    await assertRefusedSocket(5)
+    await assertServerConnections(listener, 3)
+  })
+
+  it('accepts dial of connection when connection drop listenBelow limit', async () => {
+    const listenBelow = 2
+    const closeAbove = 3
+    const port = 9900
+
+    const trasnport = tcp({ closeServerOnMaxConnections: { listenBelow, closeAbove } })()
+
+    const upgrader = mockUpgrader({
+      events: new EventEmitter()
+    })
+    const listener = trasnport.createListener({ upgrader }) as TCPListener
+    // eslint-disable-next-line @typescript-eslint/promise-function-async
+    afterEachCallbacks.push(() => listener.close())
+    await listener.listen(multiaddr(`/ip4/127.0.0.1/tcp/${port}`))
+    const { assertConnectedSocket } = buildSocketAssertions(port, afterEachCallbacks)
+
+    const socket1 = await assertConnectedSocket(1)
+    const socket2 = await assertConnectedSocket(2)
+    await assertConnectedSocket(3)
+    await assertServerConnections(listener, 3)
+
+    // Destroy sockets to be have connections < listenBelow
+    socket1.destroy()
+    socket2.destroy()
+    // After destroying 2 sockets connections will be below "listenBelow" limit
+    await assertServerConnections(listener, 1)
+
+    // Now it should be able to accept new connections
+    await assertConnectedSocket(4)
+    await assertConnectedSocket(5)
+
+    // 2 connections dropped and 2 new connections accepted
+    await assertServerConnections(listener, 3)
+  })
+
+  it('should not emit "close" event when server is stopped due to "closeAbove" limit', async () => {
+    const listenBelow = 2
+    const closeAbove = 3
+    const port = 9900
+
+    const trasnport = tcp({ closeServerOnMaxConnections: { listenBelow, closeAbove } })()
+
+    const upgrader = mockUpgrader({
+      events: new EventEmitter()
+    })
+    const listener = trasnport.createListener({ upgrader }) as TCPListener
+    // eslint-disable-next-line @typescript-eslint/promise-function-async
+    afterEachCallbacks.push(() => listener.close())
+
+    let closeEventCallCount = 0
+    listener.addEventListener('close', () => {
+      closeEventCallCount += 1
+    })
+
+    await listener.listen(multiaddr(`/ip4/127.0.0.1/tcp/${port}`))
+    const { assertConnectedSocket } = buildSocketAssertions(port, afterEachCallbacks)
+
+    await assertConnectedSocket(1)
+    await assertConnectedSocket(2)
+    await assertConnectedSocket(3)
+    await assertServerConnections(listener, 3)
+
+    // Limit reached, server should be closed but should not emit "close" event
+    expect(closeEventCallCount).equals(0)
+  })
+
+  it('should not emit "listening" event when server is resumed due to "listenBelow" limit', async () => {
```

Comment:
This test is a bit misleading, if we wanted to ensure that a listening event is not emitted when the server resumes, wouldn't we allow the server to resume and ensure that the `listeningEventCallCount` remains unchanged? 

Here that event is being incremented twice in this test

Reply:
Standard copy paste error. I wanted to "should emit" in this test case.